### PR TITLE
feat: use Prometheus to track and expose operation and API request counts

### PR DESCRIPTION
**Pull request by fiam**
This PR adds support for exposing metrics using Prometheus, which can be optionally disabled at runtime. The following metrics are exposed:

- For operations:
  - Incoming HTTP request duration
  - Incoming HTTP request count (with status code)
  - Incoming HTTP request size
  - Outgoing HTTP response size

- For upstreams:
  - Number of requests sent per host (with method and status code)
  - Duration of requests sent per host

Sample of the metrics:

```
# HELP wunderctl_api_transport_outgoing_http_request_duration Duration of outgoing HTTP requests sent to upstream APIs
# TYPE wunderctl_api_transport_outgoing_http_request_duration summary
wunderctl_api_transport_outgoing_http_request_duration_sum{host="countries.trevorblades.com",method="POST",statusCode="200"} 0.069460958
wunderctl_api_transport_outgoing_http_request_duration_count{host="countries.trevorblades.com",method="POST",statusCode="200"} 1
wunderctl_api_transport_outgoing_http_request_duration_sum{host="weather-api.wundergraph.com",method="POST",statusCode="200"} 2.718797791
wunderctl_api_transport_outgoing_http_request_duration_count{host="weather-api.wundergraph.com",method="POST",statusCode="200"} 1
# HELP wunderctl_api_transport_outgoing_http_requests_total Outgoing HTTP requests sent to upstream APIs
# TYPE wunderctl_api_transport_outgoing_http_requests_total counter
wunderctl_api_transport_outgoing_http_requests_total{host="countries.trevorblades.com",method="POST",statusCode="200"} 1
wunderctl_api_transport_outgoing_http_requests_total{host="weather-api.wundergraph.com",method="POST",statusCode="200"} 1
# HELP wunderctl_http_request_duration_seconds Duration of HTTP requests
# TYPE wunderctl_http_request_duration_seconds summary
wunderctl_http_request_duration_seconds_sum{operationName="CountryWeather",operationType="QUERY"} 2.78903725
wunderctl_http_request_duration_seconds_count{operationName="CountryWeather",operationType="QUERY"} 1
# HELP wunderctl_http_request_size_bytes Size of received HTTP requests
# TYPE wunderctl_http_request_size_bytes summary
wunderctl_http_request_size_bytes_sum{operationName="CountryWeather",operationType="QUERY"} 172
wunderctl_http_request_size_bytes_count{operationName="CountryWeather",operationType="QUERY"} 1
# HELP wunderctl_http_requests_total Total number of served HTTP requests
# TYPE wunderctl_http_requests_total counter
wunderctl_http_requests_total{operationName="CountryWeather",operationType="QUERY",statusCode="200"} 1
# HELP wunderctl_http_response_size_bytes Size of served HTTP responses
# TYPE wunderctl_http_response_size_bytes summary
wunderctl_http_response_size_bytes_sum{operationName="CountryWeather",operationType="QUERY"} 55
wunderctl_http_response_size_bytes_count{operationName="CountryWeather",operationType="QUERY"} 1
```

## Motivation and Context

Exposing metrics with Prometheus allows better monitoring of WunderGraph nodes

## Marketing Changelog

WunderGraph now supports exposing metrics about operations and upstreams using Prometheus, allowing you to monitor and understand your gateway and APIs better.

#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by jensneuse — APPROVED** [review]
I suggest adding at least one test to cover the functionality somehow.

File: packages/protobuf/wgpb/wundernode_config.ts [hunk 1]
```diff
@@ -1135,6 +1135,10 @@ export interface NodeLogging {
   level: ConfigurationVariable | undefined;
 }
 
+export interface PrometheusOptions {
```
Comment (StarpTech) [inline packages/protobuf/wgpb/wundernode_config.ts:1138]: I'm missing `enabled` flag to enable or disable the feature.

File: pkg/apihandler/metrics.go [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/wundergraph/wundergraph/pkg/metrics"
+	"github.com/wundergraph/wundergraph/pkg/wgpb"
+)
+
+const (
+	metricsNamespace = "wunderctl"
```
Comment (StarpTech) [inline pkg/apihandler/metrics.go:18]: `wunderctl` is a bit misleading. let's call it `wundernode`.

File: pkg/apihandler/metrics.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	}
+	s += len(r.Host)
+
+	if r.ContentLength != -1 {
+		s += int(r.ContentLength)
+	}
+	return s
+}
+
+func (h *operationMetricsHandler) Handler(operationHandler http.Handler) http.Handler {
```
Comment (StarpTech) [inline pkg/apihandler/metrics.go:124]: Please add some tests for this handler.

File: pkg/metrics/prometheus.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+		}
+	} else {
+		m.collectors = append(m.collectors, counterVec)
+	}
+	return &prometheusCounterVec{
+		counterVec: counterVec,
+	}
+}
+
+func (m *prometheusMetrics) NewGaugeVec(opts MetricOpts, labelNames ...string) GaugeVec {
```
Comment (StarpTech) [inline pkg/metrics/prometheus.go:86]: Please also add some tests. See https://github.com/zeromicro/go-zero/blob/master/core/metric/counter_test.go

File: pkg/metrics/prometheus.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+func (c *prometheusCounterVec) Inc(labels ...string) {
+	c.counterVec.WithLabelValues(labels...).Inc()
+}
+
+func (c *prometheusCounterVec) Add(v float64, labels ...string) {
+	c.counterVec.WithLabelValues(labels...).Add(v)
+}
+
+type prometheusGaugeVec struct {
```
Comment (StarpTech) [inline pkg/metrics/prometheus.go:25]: Please put the types in different files. counter.go, histogram.go... This is easier to grasp.

File: pkg/apihandler/metrics.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+		// XXX: Must match the number of labels used at creation
+		h.counter.Inc(strconv.Itoa(ww.statusCode))
+		h.requestDuration.Observe(elapsed.Seconds())
+		h.requestSize.Observe(float64(h.estimateRequestSize(r)))
+		h.responseSize.Observe(float64(ww.written))
+	})
+}
+
+// NewRequestMetrics returns a new RequestMetrics which can count requests.
+// XXX: Do NOT create more than one of these because the metric names will conflict.
```
Comment (StarpTech) [inline pkg/apihandler/metrics.go:142]: I don't understand.

File: pkg/apihandler/metrics.go [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+type outgoingRequestCounter struct {
+	requestsCounter metrics.CounterVec
+	requestDuration metrics.SummaryVec
+}
+
+func newOutgoingRequestCounter(m metrics.Metrics) *outgoingRequestCounter {
+	const (
+		subsystem = "api_transport"
```
Comment (StarpTech) [inline pkg/apihandler/metrics.go:192]: I also need to choose names for these things in otel. What is better `subsystem` or `component`?

File: pkg/apihandler/metrics.go [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+		h.counter.Inc(strconv.Itoa(ww.statusCode))
+		h.requestDuration.Observe(elapsed.Seconds())
+		h.requestSize.Observe(float64(h.estimateRequestSize(r)))
+		h.responseSize.Observe(float64(ww.written))
+	})
+}
+
+// NewRequestMetrics returns a new RequestMetrics which can count requests.
+// XXX: Do NOT create more than one of these because the metric names will conflict.
+func newOperationMetrics(m metrics.Metrics, operation *wgpb.Operation) operationMetrics {
```
Comment (StarpTech) [inline pkg/apihandler/metrics.go:143]: Let's not couple to the protobuf config. Let the module be standalone.

File: pkg/apihandler/metrics.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+	requestDuration metrics.SummaryVec
+	requestSize     metrics.SummaryVec
+	responseSize    metrics.SummaryVec
+}
+
+func (h *operationMetricsHandler) inc(statusCode int) {
+	h.counter.Inc(strconv.Itoa(statusCode))
+}
+
+func (h *operationMetricsHandler) estimateRequestSize(r *http.Request) int {
```
Comment (StarpTech) [inline pkg/apihandler/metrics.go:108]: The implementation does not follow any standard. You can rely on content-length but counting header length is odd.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: docs-website/src/pages/docs/guides/monitoring-and-observability.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+a custom port or to disable it, use the `prometheus` field when calling `configureWunderGraphApplication()`:
+
+```typescript
+// wundergraph.config.ts
+
+configureWunderGraphApplication({
+    ...
+    options: {
+        prometheus: {
+            port: 8881, // Set to zero to disable
```
Comment (ggeijteman-wg) [inline docs-website/src/pages/docs/guides/monitoring-and-observability.md:31]: is that:

`*nothing*`, `port: zero,`, `port: ,` or something else :D?

Comment (StarpTech) [reply docs-website/src/pages/docs/guides/monitoring-and-observability.md:31]: We should introduce `enabled` flag. The default behavior of a zero port is random on most systems.

Comment (ggeijteman-wg) [reply docs-website/src/pages/docs/guides/monitoring-and-observability.md:31]: Honestly, I would go for a `enable` field, that seems to be the more common pattern. Hinging this on the availability of the port field will only cause trouble when toggling on/off for testing.

File: packages/protobuf/wgpb/wundernode_config.ts
Comment (StarpTech) [inline packages/protobuf/wgpb/wundernode_config.ts:1]: You forgot to introduce new `WG_` envs and update the docs.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/apihandler/metrics.go (hunk 6 above)
Comment (fiam) [reply pkg/apihandler/metrics.go:142]: It's outdated, I figured a way to be able to instant them multiple times. I've removed it.

File: docs-website/src/pages/docs/guides/monitoring-and-observability.md (hunk 10 above)
Comment (fiam) [reply docs-website/src/pages/docs/guides/monitoring-and-observability.md:31]: I've added an `enabled` option

File: packages/protobuf/wgpb/wundernode_config.ts
Comment (fiam) [reply packages/protobuf/wgpb/wundernode_config.ts:1]: Added them to the docs

File: packages/protobuf/wgpb/wundernode_config.ts (hunk 1 above)
Comment (fiam) [reply packages/protobuf/wgpb/wundernode_config.ts:1138]: I've added it

File: pkg/apihandler/metrics.go (hunk 2 above)
Comment (fiam) [reply pkg/apihandler/metrics.go:18]: Renamed

File: pkg/metrics/prometheus.go (hunk 5 above)
Comment (fiam) [reply pkg/metrics/prometheus.go]: I moved them to separate files. They ended up being < 10 lines each, though 😅

File: pkg/metrics/prometheus.go (hunk 4 above)
Comment (fiam) [reply pkg/metrics/prometheus.go:42]: I've added tests

File: pkg/apihandler/metrics.go (hunk 3 above)
Comment (fiam) [reply pkg/apihandler/metrics.go:119]: Test for this added too

**Comment by fiam** [comment]
@StarpTech @ggeijteman-wg Thanks for the reviews! I've updated the PR addressing your feedback, PTAL.

File: packages/testsuite/apps/basic/prometheus.test.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+beforeAll(async () => {
+	await wg.start();
+
+	return async () => {
+		await wg.stop();
+	};
+});
+
+describe('Prometheus', () => {
+	it('should include operation metrics', async () => {
```
Comment (StarpTech) [inline packages/testsuite/apps/basic/prometheus.test.ts:17]: Nice :+1:

**Review by StarpTech — APPROVED** [review]
Great work!

**Review by ggeijteman-wg — APPROVED** [review]

**Pull request merged**